### PR TITLE
Issue #6861: Index out of bound for all-NULL case.

### LINKED ISSUE #6861: Segmentation fault on `BETWEEN`
### What happens?

The following program triggers a segmentation fault:

```
CREATE TABLE t0(c0 TIMESTAMP);

INSERT INTO t0(c0) VALUES ((DATE '1969-12-10')), ((DATE '1969-12-16')), ((DATE '1969-12-07')), ((TIMESTAMP '1969-12-09 10:08:32')), ((DATE '1969-12-30')), ((TIMESTAMP '1969-12-21 00:06:38'));

CREATE VIEW v0(c0) AS SELECT t0.c0 FROM t0;

UPDATE t0 SET c0=(NULL);

SELECT t0.c0 FROM t0, v0 WHERE (v0.c0 BETWEEN t0.c0 AND t0.c0);
```

### To Reproduce

I build DuckDB from the last version of source code, just with `make` command. I can reproduce this bug with CLI.

### OS:

ubuntu 22.04

### DuckDB Version:

commit version b8cf6a98

### DuckDB Client:

CLI

### Full Name:

Chi Zhang

### Affiliation:

Nanjing University, National University of Singapore

### Have you tried this on the latest `master` branch?

- [X] I agree

### Have you tried the steps to reproduce? Do they include all relevant data and configuration? Does the issue you report still appear there?

- [X] I agree

### PR DESCRIPTION
**Pull request by xuke-hat**
This fixes #6861

In constructor of `SBIterator`, it directly use `gss.sorted_blocks[0]`, but in all-NULL case `gss.sorted_blocks` is empty.

BTW, should we add assert in constructor of `SBIterator`? But to do this we must make `block_count` non-const to assign it after assertion.

### CODE REVIEW

File: test/sql/join/iejoin/iejoin_issue_6861.test [hunk 1]
```diff
@@ -0,0 +1,21 @@
+# name: test/sql/join/iejoin/iejoin_issue_6861.test
+# description: Issue #6861: Index out of bound for all-NULL case.
+# group: [iejoin]
+
+statement ok
+PRAGMA enable_verification
+
+statement ok
+CREATE TABLE test(x INT);
+
```
Comment (hawkfish) [inline test/sql/join/iejoin/iejoin_issue_6861.test:10]: Can we make the tests more complete and distinguish between all nulls on either side. So

```sql
statement ok
CREATE TABLE test(x INT);

query II
SELECT * 
FROM test AS a, test AS b 
WHERE (a.x BETWEEN b.x AND b.x);
----

statement ok
INSERT INTO test(x) VALUES (1), (2), (3), (NULL), (NULL), (NULL);

statement ok
CREATE TABLE all_null AS SELECT * FROM test;

statement ok
UPDATE all_null SET x=(NULL);

query II
SELECT * 
FROM all_null AS a, all_null AS b 
WHERE (a.x BETWEEN b.x AND b.x);
----

query II
SELECT * 
FROM test AS a, all_null AS b 
WHERE (a.x BETWEEN b.x AND b.x);
----

query II
SELECT * 
FROM all_null AS a, test AS b 
WHERE (a.x BETWEEN b.x AND b.x);
----
```

I was concerned that `Sort(*l2)` had the same problem but this checks that it doesn't.

**Review by hawkfish — COMMENTED** [review]
Good catch! Looks good, just added some more tests.

**Comment by hawkfish** [comment]
> BTW, should we add assert in constructor of `SBIterator`? But to do this we must make `block_count` non-const to assign it after assertion.

That's a good call - and the way to do it would be to have a local static function extract the block count from the global sort and throw in there. But it bugs me that `SBIterator` does not handle the zero data case - no reason it can't really. I'm using it in another project and may fix it there.

File: test/sql/join/iejoin/iejoin_issue_6861.test (hunk 1 above)
Comment (xuke-hat) [reply test/sql/join/iejoin/iejoin_issue_6861.test:10]: Thanks for your advice! I added these tests.

If `l1` is not empty, then `l2` will be append in `AppendKey`, so check `l1` is enough.

**Comment by xuke-hat** [comment]
> > BTW, should we add assert in constructor of `SBIterator`? But to do this we must make `block_count` non-const to assign it after assertion.
> 
> That's a good call - and the way to do it would be to have a local static function extract the block count from the global sort and throw in there. But it bugs me that `SBIterator` does not handle the zero data case - no reason it can't really. I'm using it in another project and may fix it there.

I'm not sure about how to make `SBIterator` handle the zero data case correctly so I just add an assert as a temporary solution.

**Review by hawkfish — APPROVED** [review]
Thanks! LGTM.

**Pull request merged**